### PR TITLE
Call rust parser from java

### PR DESCRIPTION
**Pull request by JaroslavTulach**
This is a throw away PR. I am trying to use https://github.com/enso-org/enso/pull/3599 from Java. The goal is to document the necessary steps to build the whole system and demonstrate it can as a whole.

There is a primitive _build system_ in `LoadParser.sh` that invokes necessary commands. It is going to be replaced by proper `sbt` integration later in the real PR.

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
With 6de95fb I am able to call into native `hello` method implemented in Rust, pass there a Java string and get another string back. Printing:
```
enso$ ./LoadParser.sh 
loaded: org.enso.syntax2.parser.LoadParser
Hello Enso from Rust
Native method said: Hello, Enso!
```

**Comment by JaroslavTulach** [comment]
After [027934a](https://github.com/enso-org/enso/pull/3600/commits/027934a116ff0536be566c2c80ac083c17d7b9ba) we can call from JVM into the Rust parser and get a result back. However, when trying that with `add x y = x + y` expression the parser fails with:
```
Exception in thread "main" org.enso.syntax2.serialization.FormatException: Unknown discriminant in Tree.
        at org.enso.syntax2.Tree.deserialize(Tree.java:34)
        at org.enso.syntax2.LoadParser.main(LoadParser.java:21)

```

**Comment by JaroslavTulach** [comment]
With 3400b64 I am able to see:
```
Native method said: BodyBlock[[Line[Newline[0, 198566654, 0, 198561272, 0, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], App[App[App[App[App[Ident[Ident[false, 0, 0, 198566654, 0, -1202467584, 7, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], 0, 198566654, 0, 7, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], Ident[Ident[false, 0, 0, 198566654, 0, -1202467576, 3, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], 1, -1202467577, 1, 3, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], 0, 198566654, 0, 11, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], Ident[Ident[false, 0, 0, 198566654, 0, -1202467572, 3, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], 1, -1202467573, 1, 3, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], 0, 198566654, 0, 15, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], Ident[Ident[false, 0, 0, 198566654, 0, -1202467568, 3, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], 1, -1202467569, 1, 3, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], 0, 198566654, 0, 19, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], OprApp[Ident[Ident[false, 0, 0, 198566654, 0, -1202467564, 1, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], 0, 198566654, 0, 1, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], org.enso.syntax2.serialization.Either@8b96fde, Ident[Ident[false, 0, 0, 198566654, 0, -1202467562, 1, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], 0, -1202467562, 0, 1, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], 1, -1202467565, 1, 3, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], 0, 198566654, 0, 23, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], Ident[Ident[false, 0, 0, 198566654, 0, -1202467560, 232, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], 1, -1202467561, 1, 232, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]], 0, 198566654, 0, 256, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]]]], 0, -1202467584, 0, 256, java.nio.DirectByteBuffer[pos=0 lim=256 cap=256]]
```
in Java!

**Comment by kazcw** [comment]
There are a bunch of reasons this doesn't work. Why don't we work out an interface, and I'll implement it?

How does this look:
```java
class Parser {
    public final native Tree parse(ByteBuffer input);
}
```

The input will be interpreted as utf-8; non-utf-8 sequences will cause warnings, but we'll try to parse as much as possible.

**Comment by JaroslavTulach** [comment]
> interface, and I'll implement it?

The interface currently is:
```java
private static native ByteBuffer hello(ByteBuffer buf);
```
I will rename the method at the end, but `ByteBuffer` (of UTF-8 encoded string) in and `ByteBuffer` of `LITTLE_ENDIAN` encoded binary format to be read via `Tree.deserialize` out seems completely fine.

File: build.sbt [hunk 1]
```diff
@@ -654,6 +655,24 @@ lazy val `text-buffer` = project
...
   )
 
+val generateRustParser = TaskKey[Unit]("generateRustParser", "Generates parser sources")
+val generateRustParserSettings = generateRustParser := {
+    import sys.process._
+    Seq("cargo", "run", "-p", "enso-parser-generate-java", "--bin", "enso-parser-generate-java", "lib/rust/parser/generate-java/java/org/enso/syntax2/") !
+}
+
+lazy val `syntax-rust-definition` = project
+  .in(file("lib/rust/parser"))
```
Comment (JaroslavTulach) [inline build.sbt:665]: This definition includes #3555 generated classes into `runtime.jar`. Run `sbt buildEngineDistribution` and then:
```bash
enso$ unzip -l runtime.jar | grep org/enso/syntax2 | cut -c31-
org/enso/syntax2/
org/enso/syntax2/serialization/
org/enso/syntax2/Line.class
org/enso/syntax2/MultiSegmentAppSegment.class
org/enso/syntax2/MultipleOperatorError.class
org/enso/syntax2/OperatorBlockExpression.class
org/enso/syntax2/OperatorLine.class
org/enso/syntax2/Parser.class
...
org/enso/syntax2/serialization/Message.class
```

**Pull request closed without merging**